webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -1201,129 +1211,144 @@ final case class ContextForNoAnswer(
-              val (newFreeValues, tryPair) = findMatchingValue(report, freeValues,  (value, report) => false)
+              val (newPairedValues, pairedAgain) = findMatchingValue(report, pairedValues, unboundedVar)
 
-              logger.trace(s"found unpaired value for '${report.keyValue}'? " + tryPair)
+              logger.trace(s"Found paired again value for ${report.keyValue}? " + pairedAgain)
 
-              tryPair match {
-                case Some(v) =>
-                  recPairReports(tail, newFreeValues, v :: pairedValues, unexpected, mode)
-                case None =>
-                  // here, we don't have found any free value for that report. We are not done yet because it can be an
-                  // unexpected reports bound to an already existing value (and so the whole component should be
-                  // unexpected or if mode allows duplicates or unbound var values, it can be ok.
+              pairedAgain match {
+                case None => //really unexpected after all
+                  recPairReports(tail, newFreeValues, newPairedValues, report :: unexpected, mode)
+                case Some(v) => //found a new pair!
+                  recPairReports(tail, newFreeValues, v :: newPairedValues, unexpected, mode)
+              }
+          }
+      }
+    }
 
 
-                  val unboundedVar = (value: Value, report: ResultReports) => {
-                    mode.isSet(UnexpectedReportBehavior.UnboundVarValues) && {
-                      // this predicate is simpler: it just have to be a variable
-                      value.isVar
-                    }
-                  }
+    val componentGotAtLeastOneReport = filteredReports.nonEmpty
 
-                  val (newPairedValues, pairedAgain) = findMatchingValue(report, pairedValues, unboundedVar)
+    expectedComponent match {
+      case g: BlockExpectedReport =>
+        BlockStatusReport(g.componentName, g.reportingLogic, g.subComponents.flatMap { case e =>
+          checkExpectedComponentWithReports(e, filteredReports, noAnswerType, policyMode, unexpectedInterpretation)
+        }) :: Nil
 
-                  logger.trace(s"Found paired again value for ${report.keyValue}? " + pairedAgain)
+      case expectedComponent: ValueExpectedReport =>
 
-                  pairedAgain match {
-                    case None => //really unexpected after all
-                      recPairReports(tail, newFreeValues, newPairedValues, report :: unexpected, mode)
-                    case Some(v) => //found a new pair!
-                      recPairReports(tail, newFreeValues, v :: newPairedValues, unexpected, mode)
-                  }
-              }
-          }
+        val (matchValue, matchId) = expectedComponent.componentsValues.partitionMap{
+          case e: ExpectedValueMatch => Left(e)
+          case e: ExpectedValueId    => Right(e)
         }
 
+        // 1. start with the easy case: expected reports with a report ID
 
```

Comment:
The interesting part is below: 
- the previous matching was missing the missing (ahah) because it loops over `matched` reports, ie the one we received for the reportId. That won't tell us which one we should have received
- it was not check for the consitancy between expect component/value and what is on the report. There is no reason that a report with the correct reportId has a bad component/value, but things happen, and the test were testing that already. 
- we also assume that the number of reports received, when it's at least one, is always the correct number to behave like the old merge (for iteration I presume)